packages/dds/tree/src/feature-libraries/sequence-field/compose.ts

```diff
@@ -632,30 +567,66 @@ export class ComposeQueue<T> {
 		}
 	}
 
-	private dequeueBase(length: number = 0): ComposeMarks<T> {
-		const baseMark = this.baseMarks.dequeue();
-		return { baseMark, newMark: length > 0 ? { count: length } : undefined };
+	private dequeueBase(length: number = Infinity): ComposeMarks<T> {
+		const baseMark = this.baseMarks.dequeueUpTo(length);
+		const movedChanges = getMovedChangesFromMark(
+			this.moveEffects,
+			baseMark,
+			this.baseMarks.revision,
+		);
+
+		const newMark = createNoopMark(
+			baseMark.count,
+			movedChanges,
+			getOutputCellId(baseMark, this.baseMarks.revision, this.revisionMetadata),
+		);
+		return { baseMark, newMark };
 	}
 
-	private dequeueNew(length: number = 0): ComposeMarks<T> {
+	private dequeueNew(length: number = Infinity): ComposeMarks<T> {
+		const newMark = this.newMarks.dequeueUpTo(length);
+		const baseMark = createNoopMark(
+			newMark.count,
+			undefined,
+			getInputCellId(newMark, this.newMarks.revision, this.revisionMetadata),
+		);
+
 		return {
-			baseMark: length > 0 ? { count: length } : undefined,
-			newMark: this.newMarks.dequeue(),
+			baseMark,
+			newMark,
 		};
 	}
 
 	private dequeueBoth(): ComposeMarks<T> {
+		const length = this.peekMinLength();
+		const baseMark = this.baseMarks.dequeueUpTo(length);
+		let newMark = this.newMarks.dequeueUpTo(length);
+		const movedChanges = getMovedChangesFromMark(
+			this.moveEffects,
+			baseMark,
+			this.baseMarks.revision,
+		);
+
+		if (movedChanges !== undefined) {
+			assert(newMark.changes === undefined, "Unexpected node changeset collision");
+			newMark = withNodeChange(newMark, movedChanges as T);
+		}
+
+		return {
+			baseMark,
+			newMark,
+		};
+	}
+
+	private peekMinLength(): number {
 		const baseMark = this.baseMarks.peek();
 		const newMark = this.newMarks.peek();
 		assert(
 			baseMark !== undefined && newMark !== undefined,
-			0x697 /* Cannot dequeue both unless both mark queues are non-empty */,
+			0x697 /* Cannot peek length unless both mark queues are non-empty */,
```

Comment:
This change has to go through the tooling
```suggestion
			"Cannot peek length unless both mark queues are non-empty",
```

Reply:
Thanks.